### ISSUE TITLE
Feature request: pipelines that can communicate via RAM on the same node

### ISSUE THREAD

**Issue by RaananHadar** [pachyderm/pachyderm#5120]
**What is the goal / desired outcome?**
I want a way to transfer data between pipelines in RAM on the same k8s node.

**If there is a way to accomplish this today via workaround, what does that require?**
Lets describe an example scenario:
My pipeline downloads a lot of data that does not need to be version controlled, then transforms it via a GPU.
The current way to implement this in Pachyderm:
1. Implement a small "monolith" pipeline that runs in a single container that does both the downloading and the GPU transform. If I want the GPU to be efficient, I will probably proceed to implement my own 2 stage pipeline mechanism within the container to keep all the resources busy. This forces the developer to use another pipeline mechanism that is not Pachyderm and this stage of the pipeline is a monolith, it becomes complex to maintain.
2. Have a 2 stage Pachyderm pipeline: a downloader stage followed by a GPU transformer stage. In this case, the software is more micro-service like but the data is persisted in Pachyderm and we pay for the extra IO to the object storage which is very inefficient. 
3. To achieve this today I can probably put an object storage in a ramdisk on each node, for example minio or mount a k8s volume, and have the downloader stage store the data and the GPU transformer pull the data and clean it up. But this is kind of a hack that circumvents Pachyderm's storage API.

**(Optional) What is your proposal for a feature to solve this?**
So this is a long term feature request, akin to what Spark did as an evolution from Hadoop. But I think there is a huge potential for performance gain and software maintainability here. I also assume that because RAM is limited, it is best approached [when this feature is done](https://github.com/pachyderm/pachyderm/issues/4029). I think that probably the best way to implement this feature involves extending the pipeline replication controller to include behind the scenes the containers for both pipelines stages. This will ensure that they will get scheduled on the same node and a ramdisk shared volume mount to transfer the data will help tie things together.

**Comment by jdoliner** [comment]
I understand the use case here. I'm not sure if sharing memory between pipeline steps is a viable solution though. Sharing memory between containers is a little complicated, although I think it's possible. But it would impose a lot of limitations on the design of pipelines. In particular this would only work between pipelines that are running on the same machine, which isn't generally going to be true. I think probably a more robust solution to this would be to have a shared cache of data that gets deployed on each node. That cache would hold chunks of data that were accessed or created by workers on that node, and would be used when another worker wanted to access the same data. This would give you the same benefit if pods happened to be co-located, and you could force them to co-locate if you wanted, but it wouldn't break if they weren't co-located, it would just be slower. In that solution the data would still be version controlled in Pacyhderm though, so I don't think this would satisfy your use case without us also adding data retention policies.

**Comment by RaananHadar** [comment]
I completely agree with your analysis @jdoliner .

I think it is extremely important for you not to:
1. Force the user code to implement new heuristics to deal with this situation, so in fact, the code "knows" that it is running inside Pachyderm and has to deal with a custom situation.
2. Force the data scientist writing the pipelines in Pachyderm to do some intimate k8s optimizations such as patching the pipeline with k8s [pod affinities](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/) because I definitely don't expect most Pachyderm users to have a deep understanding of k8s.

With that in mind, the simplest solution I have in mind was If I could simply deploy Pachyderm with a RAM based object storage that exists over the local nodes. This way, the solution is really elegant as you maintain the exact same Pachyderm interface. The issues with this solution, not surprisingly, is that RAM is limited:
1. You would definitely want this after data retention is available in Pachyderm for this to be viable.
2. The user would preferably want to have the ability to still use his "classic" object storage where this is a non-issue. Currently you can only deploy Pachyderm with a single object storage configuration.

So in summary, I believe that the recommended solution is to have multiple storage classes in Pachyderm. This way you can point out which storage class you want to use in your pipeline spec and combined with data retention, this will solve this quite elegantly. Mind you, as I said before, I am fine with long term feature. I just want this addressed eventually :)

**Comment by RaananHadar** [comment]
I am closing this issue, as I gave this some thought and I am suggesting #5359 as an alternative that is a lot more coherent to Pachyderm's way.
